avm/src/vm/memory.rs

```diff
@@ -207,9 +207,33 @@ impl HandlerMemory {
 
   /// Stores a nested fractal of data in a given address.
   pub fn write_fractal(self: &mut HandlerMemory, addr: i64, val: &[(usize, i64)]) {
-    let a = self.mems.len();
-    self.mems.push(val.to_vec().clone());
-    self.set_addr(addr, a, std::usize::MAX);
+    if addr >= 0 && self.addr.0.len() > (addr as usize) {
+      let a = self.addr.0[addr as usize].0;
+      let old_fractal = &self.mems[a];
+      for i in 0..old_fractal.len() {
+        if old_fractal[i].0 == self.mems.len() - 1 {
+          drop(old_fractal);
+          self.mems.pop();
+          break;
+        }
+      }
+      self.mems[a] = val.to_vec().clone();
+    } else if addr <= CLOSURE_ARG_MEM_END && self.addr.1.len() > ((addr - CLOSURE_ARG_MEM_START) as usize) {
+      let a = self.addr.1[(addr - CLOSURE_ARG_MEM_START) as usize].0;
+      let old_fractal = &self.mems[a];
+      for i in 0..old_fractal.len() {
+        if old_fractal[i].0 == self.mems.len() - 1 {
+          drop(old_fractal);
+          self.mems.pop();
+          break;
+        }
+      }
+      self.mems[a] = val.to_vec().clone();
+    } else {
+      let a = self.mems.len();
+      self.mems.push(val.to_vec().clone());
+      self.set_addr(addr, a, std::usize::MAX);
```

Comment:
I'm a big fan of keeping this contained to one method!

Reply:
Especially when this special casing should be removed in the future and shouldn't affect the public API